### DISCUSSION TITLE
certSANs and Internal, Pub IP after talosctl apply

### DISCUSSION THREAD

**Discussion by nobleess** [siderolabs/talos#9623] [Q&A] [answered]
I create VM with Talos in Cloud
My machine got two ip example (int: 1.1.1.1, pub: 2.2.2.2)

I forgot add public IP in slice machine.certSANs 
How i can path this, after talosctl apply -i

How i have
rpc error: code = Unavailable desc = connection error: desc = "transport: authentication handshake failed: tls: failed to verify certificate: x509: certificate is valid for  1.1.1.1  not 2.2.2.2"
Help)))

**Comment by smira** [comment, marked as answer]
If you can't connect over 1.1.1.1, there's no way to fix it, just re-provision the VM.

**Reply by nobleess** [reply]
How its work in awsCloud

**Reply by smira** [reply]
Talos has support for "platforms", which is the infrastructure Talos runs on.

Platforms support stuff like external IP - that is what you're looking for.

So on AWS Talos will just automatically do the right thing.

**Reply by nobleess** [reply]
controller failed {"component": "controller-runtime",               
 "controller": "k8s.KubeletStaticPodController", "error": "error refreshing pod status: error fetching pod status: an error   
 on the server (\"Authorization error (user=apiserver-kubelet-client, verb=get, resource=nodes, subresource=proxy)\") has     
 prevented the request from succeeding"}
 i have one more problem

**Reply by smira** [reply]
If you hit a problem:

1. Run `talosctl heatlh -n <controlplane node>` and look into the output.
2. Look into [Troubleshooting](https://www.talos.dev/v1.8/introduction/troubleshooting/).
3. Read [Getting started](https://www.talos.dev/v1.8/introduction/getting-started/).
4. If nothing helps, gather `talosctl support` bundle and create a separate discussion.

**Reply by nobleess** [reply]
<img width="935" alt="image" src="https://github.com/user-attachments/assets/0d18baaf-38ef-454a-bdb8-f9ded260ebc0">

**Reply by smira** [reply]
You can find yourself in the support bundle, `kube-apiserver` logs:

```
2024-11-04T12:47:10.734698001Z stderr F E1104 12:47:10.734560       1 run.go:74] "command failed" err="failed to apply admission: failed to initialize admission plugin \"PodSecurity\": PodSecurity invalid: exemptions.namespaces[1]: Duplicate value: \"kube-system\""
```

Most probably you patched machine configuration wrong way.

**Reply by nobleess** [reply]
problem in here 
<img width="568" alt="image" src="https://github.com/user-attachments/assets/a7cf956a-d159-406d-86a7-7effe2cea196">

**Reply by smira** [reply]
yes, but this is unrelated, it goes from your machine config which you modified to have this issue.

**Reply by nobleess** [reply]
Used talosctl edit mc, apply dont fix this
I wonna ctl as k9s for talosctl

**Reply by smira** [reply]
I'm sorry, I don't quite understand you. 

If you have an issue, please describe it fully.

**Reply by nobleess** [reply]
hello 
Can I configure a new node through the node entity?

**Reply by smira** [reply]
Please create a separate discussion with all the details for your question. I don't quite understand what you're asking about.

**Reply by agusdallalba** [reply]
I'm trying out Talos for the first time and I just hit this issue.

I'd say it's a bug in talosctl gen config, but it may also be a documentation issue.

I was following https://www.talos.dev/v1.10/introduction/prodnotes/, provisioning the first node using DNS for endpoint HA.

    talosctl gen config "$CLUSTER_NAME" \
        https://kube.example.com:6443 \
        --with-secrets secrets.yaml \
        --config-patch @config-patch.yaml \
        --output-dir _out

This generates a controlplane.yaml that contains:

    machine:
        certSANs: []
    cluster:
        controlPlane:
            endpoint: https://kube.example.com:6443
        apiServer:
            certSANs:
                - kube.example.com

Notice the empty `machine.certSANs`.

I applied this controlplane.yaml to the first node and everything appeared to go well.

    talosctl --insecure --nodes $IP apply-config --file _out/controlplane.yaml
    talosctl config merge _out/talosconfig
    talosctl config endpoint kube.example.com

In that last command I deviated from the documentation, because it made sense to me that the kube API endpoint would be in the same host as the talos API endpoint.

However any command after this fails with the confusing error message `error executing bootstrap: rpc error: code = Unavailable desc = connection error: desc = "transport: authentication handshake failed: tls: failed to verify certificate: x509: certificate is valid for talos-j9b-tb5, not kube.example.com"`, where `talos-j9b-tb5` is generated randomly.

Here I thought I had messed up and had to reprovision. Despite appearances this is a client error, and is possible to fix it by setting `talosctl config endpoint $IP`.

In the case of @nobleess, where `$IP` was unreachable, they probably could have salvaged the node by editing `/etc/hosts` and adding a line like `talos-j9b-tb5 $IP`.

I see two simultaneous solutions to this.

 -  It makes sense to me that the kube API endpoint would be in the same host as the talos API endpoint. Therefore **`talosctl gen config` should set `machine.certSANs[]` to the endpoint hostname in control plane nodes**, the same way it already does for `cluster.apiserver.certSANs[]`.

    In fact the `--additional-sans` flag of `talosctl gen config` (which is a cluster-scoped command, not a machine-scoped command) also includes the host in both lists:

        talosctl gen config "$CLUSTER_NAME" \
            https://ENDPOINT:6443 \
            --output-dir _out \
            --additional-sans ADDITIONAL

    Results in _out/controlplane.yaml containing:

        machine:
            certSANs:
                - ADDITIONAL
        cluster:
            controlPlane:
                endpoint: https://ENDPOINT:6443
            apiServer:
                certSANs:
                    - ENDPOINT
                    - ADDITIONAL

- As a (very useful) hack, there should be a talosctl global flag that allows you to ignore the CN/SAN of the endpoint certificate. This could be useful to recover a cluster that has been renumbered.

Kind regards.

**Reply by Iheanacho-ai** [reply]
Hi @agusdallalba 

I totally understand where you are coming from, the thing is,

Talos cares about two different APIs

- the Kubernetes API, the one you set for its HA endpoint as ( https://kube.example.com:6443)
- The Talos API, (used by talosctl to talk to the nodes)

When you ran talosctl gen config, it added kube.example.com to the Kubernetes API certs (so Kubernetes works).

But it did not add kube.example.com to Talos API certs  (machine.certSANs)

So when you told talosctl to connect to the Talos API at kube.example.com.

The Talos API cert did not recognize that name, and thus the TLS verification failed.

Thats, why step 12 of the production clusters documentation asks that you configure your endpoint: https://www.talos.dev/v1.10/introduction/prodnotes/#step-12-set-endpoints-of-your-control-plane-nodes

@smira and @rothgar do you have a different answer?

**Reply by smira** [reply]
Talos and Kubernetes APIs are different in the way HA is achieved:

* Kubernetes API is always a single endpoint, which should be backed by a proper TCP load-balancer to achieve true HA for external cluster access (in-cluster access is made HA with Talos's KubePrism)
* Talos API is HA via client-side loadbalancing and automatic failover; so our recommended path is to put IP addresses of controlplane node as endpoints.

Talos automatically manages cert SANs, but it can't know about external DNS names, so these should be added to the machine config manually.

There is no "lock out" scenario here, as IPs are always pushed to the cert SANs automatically.
